### PR TITLE
Send multiple breakpoints in one debugger request

### PR DESCRIPTION
**Pull request by JWZ2018**
Release note: none
Summary:
Previously, the DebugAdapter sends breakpoint requests one by one to Prepack. This works for the CLI because breakpoints are set one at a time. For an IDE, multiple breakpoints are sent in a single request upon startup and a single response with all the breakpoints set is expected. This PR sends allows sending multiple breakpoints in a single request to Prepack and a single response is received.

Test Plan:
Usage: `node lib/debugger/mock-ui/debugger-cli.js --adapterPath <path-to-adapter> --prepackRuntime <prepack runtime command> --sourceFile <file to prepack> --debugInFilePath <input file for debugger> --debugOutFilePath <output file for debugger>`
e.g.: `node lib/debugger/mock-ui/debugger-cli.js --adapterPath lib/debugger/adapter/DebugAdapter.js --prepackRuntime "node lib/prepack-cli.js" --sourceFile test/serializer/basic/Date.js --debugInFilePath src/debugger/.sessionlogs/engine2adapter.txt --debugOutFilePath src/debugger/.sessionlogs/adapter2engine.txt`

### CODE REVIEW

File: src/debugger/Debugger.js [hunk 1]
```diff
@@ -129,23 +129,31 @@ export class DebugServer {
     switch (command) {
       case DebugMessage.BREAKPOINT_ADD_COMMAND:
         invariant(args.kind === "breakpoint");
-        this._breakpoints.addBreakpoint(args.filePath, args.line, args.column);
-        this._channel.sendBreakpointAcknowledge(DebugMessage.BREAKPOINT_ADD_ACKNOWLEDGE, requestID, args);
+        for (let bp of args.breakpoints) {
```
Comment (yinghuitan) [inline src/debugger/Debugger.js]: I would suggest you change "BreakpointCollection" class to "BreakpointManager" and move all these breakpoint related handling logic into BreakpointManager.

File: src/debugger/channel/AdapterChannel.js [hunk 2]
```diff
@@ -51,7 +51,7 @@ export class AdapterChannel {
       this._eventEmitter.emit(DebugMessage.PREPACK_READY_RESPONSE, result);
       this.trySendNextRequest();
     } else if (messageType === DebugMessage.BREAKPOINT_ADD_ACKNOWLEDGE) {
-      let dbgResponse = this._marshaller.unmarshallBreakpointAddResponse(requestID);
+      let dbgResponse = this._marshaller.unmarshallBreakpointsAddResponse(requestID, parts.slice(2).join(" "));
```
Comment (yinghuitan) [inline src/debugger/channel/AdapterChannel.js]: I can see this is violating the design of marshaller in several aspects:
1. All the parts related join slice parsing should move inside unmarshalXXX() call. This is true not only for this change, but all the other places you called slice() method.
2. I did not see why you need to parse parts and reqeustID fields from message at the beginning of "_processPrepackMessage". Please refactor this code to move all message parsing code into unmarshalXXX() API.
3. With above suggestions, unmarshalXXX() API should only take a single message as argument. So you only need a single marshal() public API which dispatchs its call to individual private _unmarshalXXX() APIs.

File: src/debugger/channel/DebugChannel.js [hunk 3]
```diff
@@ -81,7 +82,7 @@ export class DebugChannel {
         args = runArgs;
         break;
       case DebugMessage.BREAKPOINT_ADD_COMMAND:
-        args = this._marshaller.unmarshallBreakpointArguments(requestID, parts.slice(2));
+        args = this._marshaller.unmarshallBreakpointsArguments(requestID, parts.slice(2).join(" "));
```
Comment (yinghuitan) [inline src/debugger/channel/DebugChannel.js]: My comments to _processPrepackMessage() to AdapterChannel applies here as well.

File: src/debugger/types.js [hunk 4]
```diff
@@ -30,13 +30,17 @@ export type PrepackLaunchArguments = {
...
 
-export type BreakpointArguments = {
-  kind: "breakpoint",
+export type Breakpoint = {
   filePath: string,
   line: number,
   column: number,
 };
 
+export type BreakpointsArguments = {
```
Comment (yinghuitan) [inline src/debugger/types.js]: You used to support the single breakpoint request format. Now you completely switched to multi-breakpoint requests. 
Do you need to support both? 
My understanding is that, debugger UI will send multi-breakpoint requests during launch, but will use single breakpoint request for later individual breakpoint operations so you need both.

Comment (JWZ2018) [reply src/debugger/types.js]: The protocol has breakpoint requests always send a list of breakpoints, so even with one breakpoint, the UI will send a list and this change can handle it.

File: src/debugger/channel/DebugChannel.js [hunk 5]
```diff
@@ -116,20 +63,18 @@ export class DebugChannel {
     this._requestReceived = false;
   }
 
-  sendBreakpointAcknowledge(messageType: string, requestID: number, args: BreakpointArguments): void {
-    this.writeOut(
-      this._marshaller.marshallBreakpointAcknowledge(requestID, messageType, args.filePath, args.line, args.column)
-    );
+  sendBreakpointsAcknowledge(messageType: string, requestID: number, args: BreakpointsArguments): void {
+    this.writeOut(this._marshaller.marshallBreakpointAcknowledge(requestID, messageType, args.breakpoints));
```
Comment (yinghuitan) [inline src/debugger/channel/DebugChannel.js]: Change this to use unmarshallRequest() as well.

File: src/debugger/Debugger.js [hunk 6]
```diff
@@ -129,23 +129,23 @@ export class DebugServer {
     switch (command) {
       case DebugMessage.BREAKPOINT_ADD_COMMAND:
         invariant(args.kind === "breakpoint");
-        this._breakpoints.addBreakpoint(args.filePath, args.line, args.column);
-        this._channel.sendBreakpointAcknowledge(DebugMessage.BREAKPOINT_ADD_ACKNOWLEDGE, requestID, args);
+        this._breakpoints.addBreakpointMulti(args.breakpoints);
```
Comment (yinghuitan) [inline src/debugger/Debugger.js]: If you follow the same suggestion as _processPrepackMessage() here by not splitting requestID from the request, you can collapse the remaining marshalXXX() call into a single marshal() API too. Feel free to do in later PR though.

**Pull request closed without merging**